### PR TITLE
Add first version of Spring Cache annotations

### PR DESCRIPTION
**Pull request by geoand**
~~Depends on: https://github.com/quarkusio/quarkus-spring-api/pull/12~~

This first version is intentionally limited. Spring Cache support can be greatly expanded with extra work on the caching layer once https://github.com/quarkusio/quarkus/pull/8631 is in

### CODE REVIEW

**Comment by geoand** [comment]
@gwenneg FYI

**Review by gwenneg — CHANGES_REQUESTED** [review]
Great addition to the Quarkus caching features!

I added a few suggestions and questions.

**Comment by geoand** [comment]
@gwenneg Thanks for the review. I'll address the comments tomorrow or Monday.

File: extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheProcessor.java
Comment (geoand) [reply extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheProcessor.java:40]: Yes, I plan to add support for it. It shouldn't be too tough, but it will require more control over the interceptors than we currently have. We'll need something akin to what Spring Security uses.

File: extensions/spring-cache/deployment/src/test/resources/cache-config-test.properties
Comment (geoand) [reply extensions/spring-cache/deployment/src/test/resources/cache-config-test.properties:4]: Nope, it got carried over from me copy-pasting test code. Thanks for spotting

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java:96]: It does, so I'll fix it

File: extensions/spring-cache/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (geoand) [reply extensions/spring-cache/runtime/src/main/resources/META-INF/quarkus-extension.yaml:7]: Yeah, it's fine since I'll be getting the guide in before the next release.

File: extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheUtil.java
Comment (geoand) [reply extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheUtil.java:16]: Makes sense.

File: extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheAnnotationsTransformer.java
Comment (geoand) [reply extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheAnnotationsTransformer.java:54]: Good idea, I'll add a warning

File: extensions/spring-cache/deployment/src/test/java/io/quarkus/cache/test/runtime/UnsupportedAnnotationValueTest.java
Comment (geoand) [reply extensions/spring-cache/deployment/src/test/java/io/quarkus/cache/test/runtime/UnsupportedAnnotationValueTest.java]: Fixed

**Comment by geoand** [comment]
@gwenneg PR updated per your suggestions

**Comment by geoand** [comment]
```
Quarkus - Integration Tests - OpenID Connect Adapter Code Flow FAILURE
```

is of course unrelated to this PR

File: extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheAnnotationsTransformer.java
Comment (gwenneg) [reply extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheAnnotationsTransformer.java:47]: This one doesn't look resolved.

File: extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheUtil.java
Comment (gwenneg) [inline extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheUtil.java:30]: You had an empty string check before, it is gone on purpose?

I expected `return cacheName == null || cacheName.isEmpty() ? Optional.empty() : Optional.of(cacheName);` here.

Comment (geoand) [reply extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheUtil.java:30]: Yeah I guess you are right. I should update it to be on the safe side

File: extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheAnnotationsTransformer.java
Comment (geoand) [reply extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheAnnotationsTransformer.java]: Fixed

File: extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheUtil.java
Comment (geoand) [reply extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheUtil.java]: Fixed

File: integration-tests/spring-web/src/test/java/io/quarkus/it/spring/web/SpringCacheControllerTest.java
Comment (gwenneg) [inline integration-tests/spring-web/src/test/java/io/quarkus/it/spring/web/SpringCacheControllerTest.java:28]: Isn't that risky to look for `geo` here while it could match both `geo` and `george`?

**Review by gwenneg — COMMENTED** [review]
Sorry it took me so long to come back to this PR!

I have one last question before approving it.

File: integration-tests/spring-web/src/test/java/io/quarkus/it/spring/web/SpringCacheControllerTest.java
Comment (geoand) [reply integration-tests/spring-web/src/test/java/io/quarkus/it/spring/web/SpringCacheControllerTest.java]: That's a good idea, fixed now

**Review by gwenneg — APPROVED** [review]
LGTM

I'd prefer a review from @gsmet before this is merged since this is an extension creation :)

File: extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheProcessor.java
Comment (gsmet) [inline extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheProcessor.java:38]:
```suggestion
    // some of these restrictions can probably be lifted by us doing additional work on caching after https://github.com/quarkusio/quarkus/pull/8631 lands
```

File: extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheProcessor.java
Comment (gsmet) [inline extensions/spring-cache/deployment/src/main/java/io/quarkus/spring/cache/SpringCacheProcessor.java:76]:
```suggestion
                    "Currently Spring Cache annotations can only be added to methods. Offending instance is annotation '"
```

File: extensions/spring-cache/pom.xml
Comment (gsmet) [inline extensions/spring-cache/pom.xml:14]: Not consistent with the deployment name `Spring - Cache`. Let's try to be consistent with the other Spring extensions whatever they use.

File: extensions/spring-cache/runtime/pom.xml
Comment (gsmet) [inline extensions/spring-cache/runtime/pom.xml:13]: Not consistent with the deployment name `Spring - Cache`.

File: extensions/spring-cache/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (gsmet) [inline extensions/spring-cache/runtime/src/main/resources/META-INF/quarkus-extension.yaml:6]: Maybe add `cache`?

**Review by gsmet — CHANGES_REQUESTED** [review]
OK, this is nearly ready, I caught a few minor things that needs fixing before merging.

Thanks!

**Comment by geoand** [comment]
@gsmet Comments addressed

**Review by gsmet — APPROVED** [review]
Cool, let's merge this one once CI is green.

**Comment by geoand** [comment]
This also needs to be documented before 1.5 goes out

**Pull request merged**
